### ISSUE TITLE
Feature request: Support KSP

### ISSUE THREAD

**Issue by davidjwiner** [google/dagger#2349]
For Kotlin projects, Dagger currently makes use of KAPT. This is generally quite slow since it requires generating Java stubs before the annotation processing step.

[KSP](https://github.com/google/ksp) offers an alternative to this by making it easier for annotation processors like Dagger to more directly use Kotlin's compiler plugin infrastructure. Based on initial benchmarks, using KSP can speed up build times by as much as 2x.

**Comment by Chang-Eric** [comment]
Just updating this to give a sense of where we are at. This is a priority for us that we are looking into. There are some complications though mostly due to the SPI plugins which expose the JSR-269 APIs through our public interface. There’s also the fact that our annotation processor code base is huge.

Because we still need to support our javac users, we cannot switch entirely to the KSP APIs which leaves us in a bit of a pickle since the places these are exposed are deep in our core model classes. I don’t think there is a possible way to get around this being a breaking API change, but we’ll need to figure out the right API to switch to that can support both KSP and javac SPI users.

Anyway, all this is to say, this is a high priority and we are working on this, but it is also a complicated thing to rollout for a codebase our size.

**Comment by ZacSweers** [comment]
Can you use [compiler-processing](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:room/compiler-processing/) from Room to simultaneously support both? SPI users can, for now at least, be told they're unsupported in KSP (which seems fair considering the SPI is javac-based anyway).

**Comment by Chang-Eric** [comment]
Thanks for the pointer, we are already in discussion with that team about that as a possibility =)

**Comment by CoreFloDev** [comment]
Hi any update on this? Thanks!

**Comment by Chang-Eric** [comment]
Still working on it. We're preparing a new model that will wrap both javac and KSP types like in this recent change: https://github.com/google/dagger/commit/30fa0f36a66793deb47788859a116e01e1e22bc3

**Comment by Jeffset** [comment]
> SPI users can, for now at least, be told they're unsupported in KSP

I've opened a separate issue for that to track status and discuss possible solutions as my team is concerned about KSP support in SPI.

**Comment by bcorso** [comment]
@Jeffset, once we're ready to support KSP in Dagger, we will likewise support KSP in the SPI plugins as well. I think we can probably close #2802 , as this issue should track both.

Just to give you more details on our plans for the SPI plugin:

The [new SPI model](https://github.com/google/dagger/tree/master/java/dagger/spi/model) currently exists, and we are beginning to use it internally in Dagger. However, it is not yet available for users to write their own plugins.

The new model will be a thin wrapper around the javac/ksp types. For example, [`DaggerElement`](https://github.com/google/dagger/blob/master/java/dagger/spi/model/DaggerElement.java) is a wrapper around `Element` in javac and `KSDeclaration` in KSP. If you need access to the javac or KSP types directly in your plugin, you will be able to call `.compiler()` to get which environment you're currently in (i.e. javac or KSP), and then call `.javac()` or `.ksp()` respectively to get access to the types.

Feel free to let us know if you have any concerns.

**Comment by ColtonIdle** [comment]
KSP is officially 1.0!
https://android-developers.googleblog.com/2021/09/accelerated-kotlin-build-times-with.html

My last dep that uses kapt is dagger + hilt! Do we have any estimated timeline? Just curious if this is a "likely by the end of 2021" type of feature request, or a "likely by the end of 2022" type of feature request. 😅

**Comment by ZacSweers** [comment]
Dagger maintainers are pretty good about posting updates, let's trust a push system rather than nagging about something they are obviously aware of and working on.

**Comment by GuilhE** [comment]
Curiosity: Is this the first step for Dagger/Hilt become available for KMM?

**Comment by Double-O-Seven** [comment]
Any ETA on this one? :)

**Comment by Lukard** [comment]
Can we have an update on KSP support? Thanks in advance 🙇

**Comment by Chang-Eric** [comment]
Updating here since it has been some time. We've been making progress on the KSP migration, which includes making changes to Room's xprocessing libraries which we are using. Things were recently delayed for a while as the migration uncovered and exacerbated an existing issue which is the main change of the 2.41 release we just did. This issue caused a large delay not only in pulling people off the migration but also because we were afraid of making the problem worse by continuing the migration.

For that reason, it is still hard to give any sort of estimate as it is really the unknowns that we're going to hit that will take the bulk of the time. Just that one issue discovered set us back a couple months. With that said, now that the 2.41 release is finally out, we are now unblocked to continue the migration work.

**Comment by Lukard** [comment]
@Chang-Eric Thank you a lot. I really appreciate the feedback.

**Comment by osipxd** [comment]
Just click this button
<img width="316" alt="image" src="https://user-images.githubusercontent.com/3684355/164785182-aa273dd2-5763-4642-9742-47699d77041b.png">

**Comment by byates-fleetian** [comment]
KSP support is something my team urgently needs. To the point that we are considering finding a solution other than dagger.

**Comment by OussamaHaff** [comment]
@byates-fleetian, please take a look at @ZacSweers’s blog post to help keep the discussion about the actual content of the feature rather than trying to put pressure on the maintainers 🙏
(I’m not a maintainer/contributer) 

https://www.zacsweers.dev/you-are-not-blocked/

**Comment by TWiStErRob** [comment]
Considering the number of votes on this, which I would put around 4-500, a better question might be:
"Is there any way we (the community) can help?"

**Comment by kles4enko** [comment]
@TWiStErRob As usual in the OSS community – PRs are welcome ;)

**Comment by bcorso** [comment]
**Quick Update:**

We’ve recently wrapped up the bulk of our migration to [XProcessing](https://github.com/androidx/androidx/tree/androidx-main/room/room-compiler-processing) -- a library that abstracts away the Javac APIs to allow Dagger’s processor to run in both Javac and KSP environments (see [commits here](https://github.com/google/dagger/pulls?q=is%3Apr+is%3Aclosed+XProcessing+)).

However, along the way, there was a number of places in our codebase where we had to fall back to using Javac when there was no XProcessing equivalent. Thus, we’re now working with the XProcessing team to add support for these missing features.

Unfortunately, it’s unlikely that contributing PRs will help much in this phase of the project. The bulk of the work will be dedicated to designing the APIs, so any PRs would just shift our time from designing APIs to reviewing designs for APIs. In addition, this will require some level of coordination between the XProcessing and Dagger codebases, which will be easier to coordinate from our side.

However, once the bulk of that work is done, we will need some help testing with real projects using KSP, so we’ll definitely let you all know once we get there.

**Comment by mslalith** [comment]
it's been a long time since last update. Any planned ETA?

**Comment by functionobject** [comment]
This is quite a blocker for compose.  To use 1.3.0 I need Kotlin 1.7, but that removes kapt.  I'm sure I can't be the only one.  Is a resolution of this issue close?

**Comment by larryxiao625** [comment]
这是来自QQ邮箱的自动回复邮件。你的邮件已经收到，我将在尽快给你回复。

**Comment by NitroG42** [comment]
> This is quite a blocker for compose. To use 1.3.0 I need Kotlin 1.7, but that removes kapt. I'm sure I can't be the only one. Is a resolution of this issue close?

Kapt (annotation processor) is not removed at all in Kotlin 1.7, it is the Kotlin Android Extensions that are removed (a way to use Android View directly from activity/fragment)

**Comment by zsmb13** [comment]
> in Kotlin 1.7, it is the Kotlin Android Extensions that are removed

It's besides the point here, but even those are only removed by [in 1.8] (https://android-developers.googleblog.com/2022/02/discontinuing-kotlin-synthetics-for-views.html).

**Comment by jakoss** [comment]
Even considering all previous (fair) statements that kapt is not removed (AFAIK there is no plan to remove kapt, it will go into "maintenance mode" as all ksp is a clear successor, but it is not getting remove anytime soon) - compose compiler version is not bound to other compose libraries. You can upgrade everything to 1.3.0 and leave compiler on 1.2.0. Source [here](https://android-developers.googleblog.com/2022/06/independent-versioning-of-Jetpack-Compose-libraries.html) And you can check compiler-kotlin compatibility chart [here](https://developer.android.com/jetpack/androidx/releases/compose-kotlin)

**Comment by larryxiao625** [comment]
这是来自QQ邮箱的自动回复邮件。你的邮件已经收到，我将在尽快给你回复。

**Comment by nitinsethi86** [comment]
Thanks a ton for keeping us updated. I have a question. When KSP support is added to dagger, will that also apply to dagger-android?

**Comment by Jean-Daniel** [comment]
> Thanks a ton for keeping us updated. I have a question. When KSP support is added to dagger, will that also apply to dagger-android?

dagger-android is Hilt.

>  Also note that this is only an update for Dagger (not Hilt) -- Hilt will also need to be migrated to KSP once the work on migrating Dagger stabilizes.

**Comment by TWiStErRob** [comment]
No, @Jean-Daniel, [dagger-android](https://dagger.dev/dev-guide/android) is [`com.google.dagger:dagger-android`](https://mvnrepository.com/artifact/com.google.dagger/dagger-android) with [`@ContributesAndroidInjector`](https://dagger.dev/api/latest/dagger/android/ContributesAndroidInjector.html) being the most known entry point. This predates [`com.google.dagger:hilt-android`](https://mvnrepository.com/artifact/com.google.dagger/hilt-android) by 5 years.

**Comment by nitinsethi86** [comment]
Unfortunately we still use the dagger-android artifact that @TWiStErRob mentioned. If there is any additional work required to support KSP in dagger-android, I don't expect the Guava / dagger team to work on it. I just want confirmation so that I can say that KSP is 1 more reason to move away from dagger-android.

**Comment by neworld** [comment]
Unfortunatelly our 300-fragment big project also adopted `dagger-android`. For good or for worse, we heavily depend on 
 `dagger.android` feature of components per fragment, and the hilt is totally opposite. Migration will require extreme effort and missing KSP support hardly justifies it, so I still have poor man hope to see KSP support in `dagger-android` as well :crossed_fingers:

**Comment by bcorso** [comment]
Our plan is to eventually support KSP with `dagger-android`, but it would likely come after support for Hilt.

**Comment by a-blekot** [comment]
just wondering when it will be done) maybe there is progress tracker or something like this?

**Comment by mslalith** [comment]
maybe [XProcessing progress](https://github.com/google/dagger/issues?q=XProcessing+is%3Aopen) could help

**Comment by DHosseiny** [comment]
I am wondering if dagger migrated to ksp it will support kotlin native? Or will it helps dagger to support kotlin native in the feature?

**Comment by jakoss** [comment]
I should open the door for that in the future i think, that's the biggest advantage of ksp 😁

**Comment by iamarjun** [comment]
+1

**Comment by hmzgtl16** [comment]
+1

**Comment by mslalith** [comment]
Hoping some good news this IO 🤞

**Comment by isfaaghyth** [comment]
I/O mentioned this progress, hope we could see the progress updates here and ETA!

**Comment by ColtonIdle** [comment]
ive been playing around with dagger snapshot since https://github.com/google/dagger/pull/3857 piqued my curiosity. looks like snapshot has ksp support?!

**Comment by JJSarrasin** [comment]
Sorry to bother all subscribers.

KAPT to KSP is one thing, but can we expect KMM soon after that or it is not planned yet ?

**Comment by IlyaGulya** [comment]
If a community member wants to help with KSP migration, will his pull requests be accepted?
For instance, I can take several issues from this list https://github.com/google/dagger/issues?q=XProcessing+is%3Aopen

**Comment by TWiStErRob** [comment]
@IlyaGulya there's only 3 issues there, the rest are PRs. See the "merge" vs circle icons on the left.

**Comment by Chang-Eric** [comment]
@IlyaGulya Thanks for the offer, we really appreciate it. Unfortunately, it probably isn't so easy for community members to contribute to the KSP migration. We're tracking a lot of the issues we're discovering internally because they have stacktraces/information from internal codebases. Also, a lot of issues are actually across different parts of the stack, not just in Dagger, and so a lot of the work comes from coordinating fixes across those other projects. Additionally, there are extra complications from the process to push code from our internal systems out here that just makes it difficult for us to take large code contributions from the community in general.

@bcorso is planning to update this thread soon with more details on where we're at and some of the issues we're currently working on. We know this has been going on for a long time, so we really appreciate all of your patience and support here.

**Comment by shenguojun** [comment]
Waiting for the good news, our team will move to KSP once Dagger is supported.

**Comment by nickdonnelly** [comment]
Any news here?

**Comment by ColtonIdle** [comment]
It's here!

https://github.com/google/dagger/releases/tag/dagger-2.48

**Comment by bcorso** [comment]
Hi all, Dagger 2.48 includes the alpha release of the Dagger and Hilt KSP processors. Please see the [release notes](https://github.com/google/dagger/releases/tag/dagger-2.48) for more details.

Also, as an update on https://github.com/google/dagger/issues/2349#issuecomment-1549890357,  the `@JvmSuppressWildcards` and `@JvmWildcard` annotations will still be needed in your Kotlin sources until the second part of that change is completed.  Currently, we have only implemented the first part of the proposed change (i.e. we’ve banned providing a binding with the same invariant type), as that was required for the KSP release.  
Unfortunately, we had to put the second part of that change on hold as it was not needed for the KSP release, and it will likely take a good amount of time to address some of the unexpected issues we ran into while trying to implement it.

**Comment by FunkyMuse** [comment]
> Hi all, Dagger 2.48 includes the alpha release of the Dagger and Hilt KSP processors. Please see the [release notes](https://github.com/google/dagger/releases/tag/dagger-2.48) for more details.
> 
> Also, as an update on [https://github.com/google/dagger/issues/2349#issuecomment-1549890357, ](https://github.com/google/dagger/issues/2349#issuecomment-1549890357,%C2%A0) the `@JvmSuppressWildcards` and `@JvmWildcard` annotations will still be needed in your Kotlin sources until the second part of that change is completed. Currently, we have only implemented the first part of the proposed change (i.e. we’ve banned providing a binding with the same invariant type), as that was required for the KSP release. Unfortunately, we had to put the second part of that change on hold as it was not needed for the KSP release, and it will likely take a good amount of time to address some of the unexpected issues we ran into while trying to implement it.

Thanks for the alpha release, great job.

Does it mean that the "2nd part" is gonna generate Kotlin instead of Java code?

**Comment by bcorso** [comment]
@FunkyMuse, for the "2nd part" I mean this:

> "Next, we’ll introduce another flag [...] that allows users to request a binding via any of its type’s variances, e.g. injecting `@Inject Foo<Bar>` or `@Inject Foo<? extends Bar>` would both be allowed (assuming the provided type can be assignable to them)."

Ideally, the implementation of that feature would not require generating Kotlin source, but admittedly that is part of what makes it difficult. Fwiw, I know generating Kotlin source is of some interest due to KMP; however, just to be upfront, it's not something we are currently working on at the moment.

**Comment by Hiraev** [comment]
Is KSP support still in alpha?

**Comment by bcorso** [comment]
The KSP story is kind complicated.

KSP1 is deprecated. The reason why Dagger's KSP1 support remained in alpha was to address some performance concerns that we found when rolling out KSP within Google. However it's unlikely that those issues will be fixed now that KSP1 is deprecated. 

As for KSP2, Dagger plans to support it but there's a number of known correctness issues that still prevent us from trying it out within Google. We're currently working with the KSP team to address those issues, but it will likely be a few months. Even after those correctness issues are fixed we will still need to see if there are any remaining performance issues before moving it out of alpha.

**Comment by mslalith** [comment]
If you are trying other DI libraries, try [kotlin-inject](https://github.com/evant/kotlin-inject). It's similar to Dagger

**Comment by cgruber** [comment]
@bcorso - any progress on this?

**Comment by bcorso** [comment]
> @bcorso - any progress on this?

So far, we've only tested KSP2 with a small subset of Dagger targets within Google, but there's been some progress the last few months (see [KSP fixes](https://github.com/google/ksp/issues?q=is%3Aissue%20author%3Abcorso%20KSP2) and [XProcessing fixes](https://b.corp.google.com/issues?q=componentid:413107%20author:bcorso%20KSP2)), and all of the known issues for those targets have been fixed.

The KSP2 fixes should be available in the next KSP release (2.1.10-1.0.32).

Users are free to try out KSP2 whenever they want; however, before we claim KSP2 is "officially" supported in Dagger, we still need to 1) increase our test coverage by enabling KSP2 for all relevant Dagger targets within Google, then 2) go through the same process with Hilt. However, I'm hoping we've caught most of the main issues and the remaining work will be finding the edge cases.

**Comment by alsutton** [comment]
It seems like it might be worth Updating "Alpha" to "Beta" on https://dagger.dev/dev-guide/ksp.html if the consensus is that it's usable, but may have edge case problems.

**Comment by cgruber** [comment]
I attempted to use the ksp variant in a bazel build, and the ksp environment in rules_kotlin and dagger mis-interacted, around classloading bits of com.google.guava.graph. Are you guys shading your deps properly? Might be worth a re-look, as I don't think there's any part of Guava that is part of the Dagger APIs. So Dagger shouldn't ever be trying to load an un-shaded com.google.common.* file.

**Comment by bcorso** [comment]
@cgruber Bazel hasn't really been our focus in open source so I don't even think we've published instructions or helpers to use KSP in Bazel. However, if you're interested in getting things working feel free to start a separate issue with the particular issue you hit (so that we don't pollute this issue). Include info on how you are setting up your dagger targets, e.g. are you trying to create a dagger workspace (deprecated) or are you using `maven.install` to grab the artifacts?

**Comment by cgruber** [comment]
Gotcha.

**Comment by G00fY2** [comment]
Seems like KSP team enabled KSP2 by default with latest [2.1.20-2.0.0](https://github.com/google/ksp/releases/tag/2.1.20-2.0.0). For me this seems a bit too early or badly communicated since our builds are now failing because of e.g. #4496

So probably most people need to explicitly set `ksp.useKSP2=false`.

**Comment by bcorso** [comment]
> For me this seems a bit too early or badly communicated

@G00fY2 Thanks for calling this out. I imagine KSP is kind of in a tough spot because there's no guarantees KSP1 will keep working on the latest Kotlin versions and it isn't actively maintained (i.e. there won't be any fixes if things start to break). At the very least, it seems the change is documented on the [KSP GitHub page](https://github.com/google/ksp?tab=readme-ov-file#ksp2-is-here), so if you hit an issue after upgrading KSP it will hopefully be easy to find the workaround.

> builds are now failing because of e.g. https://github.com/google/dagger/issues/4496

Which version of Dagger are you using? I tried to repro #4496 and it seems to be fixed in the latest Dagger version. If you're still having issues with the latest version can you please comment on #4496 with more info?

**Comment by G00fY2** [comment]
> Which version of Dagger are you using? I tried to repro [#4496](https://github.com/google/dagger/issues/4496) and it seems to be fixed in the latest Dagger version. If you're still having issues with the latest version can you please comment on [#4496](https://github.com/google/dagger/issues/4496) with more info?

We run into #4496 after updating from KSP 2.1.20-1.0.32 -> 2.1.20-2.0.0 while using Dagger/Hilt 2.56.1.

Today we updated from Room 2.6.1 to [2.7.0](https://developer.android.com/jetpack/androidx/releases/room#2.7.0) and now KSP2 seems to work (again using 2.1.20-2.0.0 and Dagger 2.56.1). 🥳 Probably Rooms transitive XProcessing deps fixed it. But this would mean the issues was not fixed by solely updating to Dagger 2.53.1 or higher as stated in the [comment](https://github.com/google/dagger/issues/4496#issuecomment-2790888282).

**Comment by ansman** [comment]
@G00fY2 I'm pretty sure that XProcessing is shaded, so updating Room would not affect the XProcessing in Room.

**Comment by bcorso** [comment]
@G00fY2 

> Only question I have, should we stick to ksp.useKSP2=false for now, since KSP2 is not "officially" supported or is it safe to use when KSP1 was previously used?

**tl;dr:** Feel free to stick to KSP1 if that's working for you (until it doesn't). Feel free to try KSP2 if it works for you, and if it doesn't feel free to file bugs if you have a concise repro example.

The unfortunate situation is that there's no guarantee that future changes in kotlin won't cause KSP1 to break at some point, and you would need to switch to KSP2 anyway (since even if you file a bug for KSP1 it's not going to get fixed).

While we're still actively working on getting KSP2 "officially" supported for Dagger/Hilt, the criteria is somewhat arbitrary -- it's basically having KSP2 for Dagger and Hilt mostly rolled out within Google (which is essentially our test bed). That said, getting there could require fixing a long tail of bugs that don't affect most users.

**Comment by Jean-Daniel** [comment]
> Today we updated from Room 2.6.1 to [2.7.0](https://developer.android.com/jetpack/androidx/releases/room#2.7.0) and now KSP2 seems to work (again using 2.1.20-2.0.0 and Dagger 2.56.1).

I had issue with room 2.6 and KSP 2 too. There were bug in handling suspending methods in Dao returning Void at least.
As all my other dependencies where up-to-date, I don't think Room bugs where related in any way to Dagger.

**Comment by StefMa** [comment]
We also can't use KSP `2.1.20-2.0.0` with Dagger `2.56.1`.
It raises errors like these:
```
Caused by: java.lang.IllegalArgumentException: not a valid name: UpdateLanguageActionModule_Provide$language-updater_debugFactory
        at com.squareup.javapoet.Util.checkArgument(Util.java:53)
        at com.squareup.javapoet.TypeSpec$Builder.<init>(TypeSpec.java:432)
        at com.squareup.javapoet.TypeSpec$Builder.<init>(TypeSpec.java:409)
        at com.squareup.javapoet.TypeSpec.classBuilder(TypeSpec.java:128)
        at com.squareup.javapoet.TypeSpec.classBuilder(TypeSpec.java:132)
        at dagger.internal.codegen.writing.FactoryGenerator.factoryBuilder(FactoryGenerator.java:139)
        at dagger.internal.codegen.writing.FactoryGenerator.topLevelTypes(FactoryGenerator.java:134)
        at dagger.internal.codegen.writing.FactoryGenerator.topLevelTypes(FactoryGenerator.java:103)
        at dagger.internal.codegen.base.SourceFileGenerator.generate(SourceFileGenerator.java:73)
        at dagger.internal.codegen.base.SourceFileGenerator.generate(SourceFileGenerator.java:68)
        at dagger.internal.codegen.processingstep.ModuleProcessingStep.generate(ModuleProcessingStep.java:137)
        at dagger.internal.codegen.processingstep.ModuleProcessingStep.generateForMethodsIn(ModuleProcessingStep.java:121)
        at dagger.internal.codegen.processingstep.ModuleProcessingStep.process(ModuleProcessingStep.java:109)
        at dagger.internal.codegen.processingstep.ModuleProcessingStep.process(ModuleProcessingStep.java:48)
        at dagger.internal.codegen.processingstep.TypeCheckingProcessingStep.lambda$process$0(TypeCheckingProcessingStep.java:96)
        at com.google.common.collect.SingletonImmutableBiMap.forEach(SingletonImmutableBiMap.java:70)
        at dagger.internal.codegen.processingstep.TypeCheckingProcessingStep.process(TypeCheckingProcessingStep.java:74)
        at dagger.internal.codegen.processingstep.ModuleProcessingStep.process(ModuleProcessingStep.java:90)
        at dagger.internal.codegen.processingstep.ModuleProcessingStep.process(ModuleProcessingStep.java:48)
        at dagger.spi.internal.shaded.androidx.room.compiler.processing.XProcessingStep.process(XProcessingStep.kt:57)
        at dagger.spi.internal.shaded.androidx.room.compiler.processing.CommonProcessorDelegate.processRound(XBasicAnnotationProcessor.kt:134)
        at dagger.spi.internal.shaded.androidx.room.compiler.processing.ksp.KspBasicAnnotationProcessor.process(KspBasicAnnotationProcessor.kt:62)
        at com.google.devtools.ksp.impl.KotlinSymbolProcessing$execute$1$1.invoke(KotlinSymbolProcessing.kt:579)
        at com.google.devtools.ksp.impl.KotlinSymbolProcessing$execute$1$1.invoke(KotlinSymbolProcessing.kt:577)
        at ksp.com.google.devtools.ksp.common.IncrementalContextBase.closeFilesOnException(IncrementalContextBase.kt:400)
        at com.google.devtools.ksp.impl.KotlinSymbolProcessing.execute(KotlinSymbolProcessing.kt:577)
        at com.google.devtools.ksp.impl.KSPLoader$Companion.loadAndRunKSP(KSPLoader.kt:37)
        at com.google.devtools.ksp.impl.KSPLoader.loadAndRunKSP(KSPLoader.kt)
        at jdk.internal.reflect.GeneratedMethodAccessor680.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at com.google.devtools.ksp.gradle.KspAAWorkerAction.execute(KspAATask.kt:580)
        at org.gradle.workers.internal.DefaultWorkerServer.execute(DefaultWorkerServer.java:63)
        at org.gradle.workers.internal.NoIsolationWorkerFactory$1$1.create(NoIsolationWorkerFactory.java:66)
        at org.gradle.workers.internal.NoIsolationWorkerFactory$1$1.create(NoIsolationWorkerFactory.java:62)
        at org.gradle.internal.classloader.ClassLoaderUtils.executeInClassloader(ClassLoaderUtils.java:100)
        at org.gradle.workers.internal.NoIsolationWorkerFactory$1.lambda$execute$0(NoIsolationWorkerFactory.java:62)
        at org.gradle.workers.internal.AbstractWorker$1.call(AbstractWorker.java:44)
        at org.gradle.workers.internal.AbstractWorker$1.call(AbstractWorker.java:41)
// Some other, probably unrelated org.gradle* stuff :) 
```

This is how our Module look like:
```kotlin
@Module
public class UpdateLanguageActionModule {
    @Provides
    internal fun provide(userAuthRepository: UserAuthRepository, someService: someService): UpdateLanguageAction =
        DefaultUpdateLanguageAction(userAuthRepository, someService)
}
```

I don't know if this is an issue on Dagger side or KSP side 🤔 
I also comment this there: https://github.com/google/ksp/issues/2407#issuecomment-2808598192

Let me know if this is related to the current discussions, if this is unrelated to dagger or if I should raise another issue because of this...

**Comment by dawidhyzy** [comment]
Hi @StefMa! 
We had the same issue. I figured out that the problem was in the generated factory name. In your case, `UpdateLanguageActionModule_Provide$language-updater_debugFactory` contains `-`, which is forbidden in character for a class name, and I assume comes from the `language-updater` project module name. I would recommend you rename it until this is fixed.

**Comment by StefMa** [comment]
@dawidhyzy thanks for the response!
You mean the Gradle Module/Project right? 🤔 This is indeed named language-updater.

Do you know who is responsible for the generated class? 🤔
I ask because renaming is not a good way for us. Naming Gradle Projects with dashes is common and nothing special. And we have a lot of those. Messing up with new names for everything isn't the way to go.
So which project should tackle the renaming to something else? 🤔 Is it KSP or Dagger?

**Comment by dawidhyzy** [comment]
@StefMa, exactly, I meant Gradle Module.

I understand. For us, it was the app module suffix that we should remove after moving it to a separate repository, so it was a good change.

Unfortunately, I don't know where this has to be fixed. I think it would be best to check with contributors(@bcorso and @Chang-Eric should be in this thread).

**Comment by alsutton** [comment]
Dashes in project names is recommended by the folk at Gradle, so there should be a test for it somewhere;  https://docs.gradle.org/current/userguide/multi_project_builds.html#sec:naming_recommendations

**Comment by bcorso** [comment]
@StefMa see https://github.com/google/dagger/issues/4680.

**Comment by ychescale9** [comment]
>The unfortunate situation is that there's no guarantee that future changes in kotlin won't cause KSP1 to break at some point, and you would need to switch to KSP2 anyway (since even if you file a bug for KSP1 it's not going to get fixed).

Unfortunately this seems to have started happening with Kotlin 2.2.0-Beta1. Enabling KSP2 is the right solution for first party tools such as Dagger and room where KSP2 support is a priority, but this isn't feasible if the project uses other KSP processors that haven't (in some cases will never) supported KSP2. I feel like this might start blocking many companies from upgrading Kotlin.

**Comment by ting-yuan** [comment]
> but this isn't feasible if the project uses other KSP processors that haven't (in some cases will never) supported KSP2.

Although there are some API changes between KSP1 and 2, would you mind sharing more details on what prevent them from working in KSP2? There might be some actionable items in KSP side.

Btw, in cases where processors aren't used in the same sources, switchng between KSP1 and KSP2 can be done in a per-gradle-project / build.gradle.kts basis in `ksp { }`.

**Comment by ychescale9** [comment]
> Although there are some API changes between KSP1 and 2, would you mind sharing more details on what prevent them from working in KSP2? There might be some actionable items in KSP side.

Here's one example: https://github.com/airbnb/epoxy/issues/1388, though not knowing about the implementation details of these 3rd party KSP processors that fail the build it can be hard to know whether the issues can be worked around in KSP 2 itself.

**Comment by dk-3cx** [comment]
Dagger's KSP support has been documented as alpha since Dagger 2.48 ([released 2023-08-30](https://github.com/google/dagger/releases?page=3#release-dagger-2.48)) — three years as of 2026-09-01. It is still labelled alpha in [the current docs](https://dagger.dev/dev-guide/ksp), including as of Dagger 2.60.1 (2026-07-06).

Meanwhile AGP has moved decisively [against **kapt**](https://developer.android.com/agents/skills/build-system/agp/agp-9-upgrade/references/ksp-kapt). AGP 9.0 enables built-in Kotlin by default, which is incompatible with `kotlin-kapt`; the sanctioned fallback is `com.android.legacy-kapt`, and the `android.builtInKotlin=false` opt-out is [scheduled for removal](https://developer.android.com/build/migrate-to-built-in-kotlin#opt-out-of-built-in-kotlin) in AGP 10.

Staying on the older toolchain is not a realistic option for us: a growing number of core Android libraries now require recent AGP and Kotlin releases, so we cannot remain indefinitely in the configuration where **kapt** was still a supported choice.

This leaves Dagger users with no non-alpha, forward-compatible path. We are requesting an official statement on the KSP stability roadmap.

### Questions

1. Is there a target release or timeframe for Dagger KSP to reach beta, and subsequently stable?
2. What specific criteria or known blockers remain before beta? Is there a checklist? We would also like to understand the concrete risk of adopting KSP now.
3. Given AGP 9 already classifies kapt as legacy and AGP 10 removes the built-in-Kotlin opt-out, is Dagger KSP expected to reach stable before AGP 10 ships?
4. Is `com.android.legacy-kapt` + Dagger-on-kapt a configuration that will remain supported through AGP 10, or should production applications plan on adopting alpha-stage KSP support instead?

A published roadmap — even an approximate one — would let teams decide between adopting alpha KSP, remaining on legacy-kapt, or migrating away from Dagger before AGP 10.

Thanks in advance for any clarity you can share.

**Comment by lukas1** [comment]
> KSP is officially 1.0! https://android-developers.googleblog.com/2021/09/accelerated-kotlin-build-times-with.html
> 
> My last dep that uses kapt is dagger + hilt! Do we have any estimated timeline? Just curious if this is a "likely by the end of 2021" type of feature request, or a "likely by the end of 2022" type of feature request. 😅

It is "Not finished by September 2026 kind of feature request". 😅

**Comment by bcorso** [comment]
@dk-3cx, sorry for the delayed response (I was OOO)! 

We dropped the ball on making an official announcement, but the good news is that Dagger KSP is considered stable as of version 2.60+ and KSP 2.3.9+. In particular, we hit our Beta criteria back in December 2025 and our Stable criteria in July 2026. You can see more details on these decisions below, and we will update the documentation to reflect this.

To explicitly answer your questions:

1. **Target for beta/stable:** Dagger KSP is officially considered stable as of version 2.60.  
2. **Criteria/blockers before beta:** There are no remaining blockers. The risk of adopting KSP now is very low.  
3. **Stable before AGP 10:** Yes, it is stable now.  
4. **Legacy-kapt vs KSP:** As for whether legacy-kapt will remain supported in AGP 10, that’s something that you’ll need to ask the AGP team. Dagger will continue to support both Javac/KAPT and KSP annotation processing, and you are free to choose whichever one works best for your application.

For the long answer, you can check out the timeline below. This has been a very long, messy road and admittedly we haven’t done a great job of communicating. In addition, the goal posts were fuzzy and kept shifting. I’ve tried to clarify as much as possible below to give an idea of why this has taken so long, and to justify the Alpha, Beta, and Stable dates I mentioned above. (Also keep in mind that while KSP was one of our top priorities, there were occasionally higher priority issues \-- not mentioned in this timeline \-- that put this work on hold for various amounts of time):

* **Early 2022** ([https://github.com/google/dagger/issues/2349\#issuecomment-1042423111](https://github.com/google/dagger/issues/2349#issuecomment-1042423111)), work started on Dagger’s migration to KSP1 using [XProcessing](https://github.com/androidx/androidx/tree/androidx-main/room3/room3-compiler-processing) (a wrapper around the annotation processor layer that supports both Javac/KAPT and KSP).  
* **May 2022** ([https://github.com/google/dagger/issues/2349\#issuecomment-1143028539](https://github.com/google/dagger/issues/2349#issuecomment-1143028539)), wrapped up our initial migration to XProcessing, though in the process we found many APIs that were still missing. We began working with the XProcessing team to design and implement these new APIs.  
* **Nov 2022** ([https://github.com/google/dagger/issues/2349\#issuecomment-1315862428](https://github.com/google/dagger/issues/2349#issuecomment-1315862428)), Dagger KSP was officially enabled and working in all of Dagger’s existing Java source tests. We began work to add Kotlin source tests and rollout Dagger KSP to real projects in Google to test correctness.  
* **May 2023** ([https://github.com/google/dagger/issues/2349\#issuecomment-1549890357](https://github.com/google/dagger/issues/2349#issuecomment-1549890357)), the rollout to real apps in Google uncovered a number of issues \-- the largest being translating Kotlin types to their Java equivalent with the correct variance. In particular, Dagger relies heavily on matching the exact type of a request (e..g Foo\<Bar\> \!= Foo\<? extends Bar\>) and for backwards compatibility, this process had to be reverse engineered to exactly match the types produced by KAPT (see [KSTypeVarianceResolver](https://github.com/androidx/androidx/blob/androidx-main/room3/room3-compiler-processing/src/main/java/androidx/room3/compiler/processing/ksp/KSTypeVarianceResolver.kt#L32-L47)). We spent the next few months tracking down and fixing all of these issues.  
* **Aug 2023** ([https://github.com/google/dagger/issues/2349\#issuecomment-1699569360](https://github.com/google/dagger/issues/2349#issuecomment-1699569360)), Dagger KSP was released as Alpha (for KSP1). At this point the main remaining issue was performance. The plan was to give a short time in Alpha for external developers to try it out and then quickly move to Beta/Stable while investigating performance.  
* **Dec 2023** ([https://android-developers.googleblog.com/2023/12/ksp2-preview-kotlin-k2-standalone.html](https://android-developers.googleblog.com/2023/12/ksp2-preview-kotlin-k2-standalone.html)), KSP2 was announced. This essentially halted our plans to move KSP1 to Beta/Stable, and we shifted our priority to supporting KSP2.  
* **Sep 2024** ([https://github.com/google/dagger/issues/2349\#issuecomment-2329852274](https://github.com/google/dagger/issues/2349#issuecomment-2329852274)), at this point we were mostly just filing issues to the KSP team trying to get KSP2 working with Dagger’s and XProcessing’s existing test suites (which were previously working with KSP1). Unfortunately, this process just took a long time. The KSP team was essentially building the KSP2 architecture from scratch, which was a massive undertaking that constrained their bandwidth to fix all of these edge cases.  
* **March 2025** ([https://github.com/google/dagger/issues/2349\#issuecomment-2725553921](https://github.com/google/dagger/issues/2349#issuecomment-2725553921)), all of the known KSP2 correctness issues were fixed, and **we were essentially back at the Alpha stage for KSP2** and ready for external users to try it out. Similar to before, the plan was to give a short time in Alpha for external developers to try it out and then quickly move to Beta/Stable while investigating performance.&nbsp;  
* **Dec 2025**, we had rolled out KSP2 to \~40% of apps in Google, and received very few reports of issues from external users, so we were pretty confident in the correctness of KSP2. **We should have officially moved Dagger KSP to Beta at this time**. The main issues remaining were just related to performance, especially cases where Dagger KSP2 was actually slower than Dagger KAPT. We decided to dedicate resources from the Dagger team to investigate performance across the various layers in the stack, including Dagger, XProcessing, KSP, and Kotlinc.&nbsp;  
* **July 2026**, we wrapped up a number of performance improvements and validated that Dagger KSP2 was faster than KSP1/KAPT. **We should have officially moved Dagger KSP to Stable at this time**.

I hope this helps clarify things, and feel free to let me know if there’s anything I haven’t answered or remains unclear.

**Comment by jbduncan** [comment]
@bcorso Yes, this does clarify things and, although I can't speak for everyone, any concerns I may have had about timelines and the messy communication have completely melted away. So, thank you for taking the time to explain everything in such great detail! 😄👍

**Comment by jbduncan** [comment]
Also, well done to the team for taking on such a huge, thankless and meandering task, your efforts have not gone unnoticed. 🎉
